switchlink/switchlink_route_test.cc

```diff
@@ -0,0 +1,738 @@
+ */
+int switchlink_create_ecmp(switchlink_db_ecmp_info_t *ecmp_info) {
+  return 0; 
+}
+
+/*
+ * Dummy function for switchlink_db_add_ecmp().
+ */
+switchlink_db_status_t
+switchlink_db_add_ecmp(switchlink_db_ecmp_info_t *ecmp_info) {
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Dummy function for switchlink_db_update_nexthop_using_by().
+ */
+switchlink_db_status_t switchlink_db_update_nexthop_using_by(
+    switchlink_db_nexthop_info_t *nexthop_info) {
+  struct test_results temp = {0};
+  if (nexthop_info) {
+    temp.nexthop_info = *nexthop_info;
+  }
+  results.push_back(temp);
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Dummy function for switchlink_db_add_nexthop().
+ */
+switchlink_db_status_t
+switchlink_db_add_nexthop(switchlink_db_nexthop_info_t *nexthop_info) {
+  struct test_results temp = {0};
+  if (nexthop_info) {
+    temp.nexthop_info = *nexthop_info;
+  }
+  results.push_back(temp);
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Dummy function for switchlink_db_get_ecmp_info().
+ */
+switchlink_db_status_t
+switchlink_db_get_ecmp_info(switchlink_db_ecmp_info_t *ecmp_info) {
+  ecmp_info->ecmp_h = TEST_ECMP_H;
+  if (ecmp_info) {
+    results[0].ecmp_info = *ecmp_info;
+  }
+  return SWITCHLINK_DB_STATUS_SUCCESS;
+}
+
+/*
+ * Dummy function for switchlink_db_get_nexthop_info().
+ * Returns success for intf handle: TEST_INTF_H1
+ * Returns failure for intf handle: TEST_INTF_H2
+ */
+switchlink_db_status_t
+switchlink_db_get_nexthop_info(switchlink_db_nexthop_info_t *nexthop_info) {
+
+  if (nexthop_info->intf_h == TEST_INTF_H1) {
+    nexthop_info->nhop_h = 1;
+    return SWITCHLINK_DB_STATUS_SUCCESS;
+  } else if (nexthop_info->intf_h == TEST_INTF_H2) {
+    nexthop_info->nhop_h = 2;
+    return SWITCHLINK_DB_STATUS_ITEM_NOT_FOUND;
+  }
+
+  return SWITCHLINK_DB_STATUS_ITEM_NOT_FOUND;
+}
+
+/*
+ * Test fixture.
+ */
+class SwitchlinkRouteTest : public ::testing::Test {
+protected:
+  struct nl_msg *nlmsg_ = nullptr;
+
+  // Sets up the test fixture.
+  void SetUp() override { ResetVariables(); }
+
+  // Tears down the test fixture.
+  void TearDown() override {
+    if (nlmsg_) {
+      nlmsg_free(nlmsg_);
+      nlmsg_ = nullptr;
+    }
+  }
+
+  void ResetVariables() {
+    // result variables
+    results.clear();
+  }
+};
+
+/*
+ * Creates an IPv4 route
+ *
+ * Validates switchlink_process_route_msg(). It parses an
+ * RTM_NEWROUTE message, invokes switchlink_create_route()
+ * for creating route entry.
+ *
+ */
+TEST_F(SwitchlinkRouteTest, createIPv4Route) {
+  struct rtmsg hdr = {
+      .rtm_family = AF_INET,
+      .rtm_dst_len = 24,
+      .rtm_src_len = 24,
+      .rtm_tos = 1,
+      .rtm_table = 1,
+      .rtm_protocol = 1,
+      .rtm_scope = 1,
+      .rtm_type = 1,
+      .rtm_flags = 0x1,
+  };
+
+  const uint32_t src_addr = IPV4_ADDR(10, 10, 10, 10);
+  const uint32_t dst_addr = IPV4_ADDR(20, 20, 20, 20);
+  const uint32_t gateway_addr = IPV4_ADDR(20, 20, 20, 1);
```

Comment:
Should we test for invalid IP addresses?

Reply:
yes, we can for making it more robust. But we don't have any error checking mechanism in the source code for checking invalid ip addresses.

Reply:
@5abeel 
The Unit Under Test (UUT) parses fields in RFC 3594 Netlink messages and calls external functions to process the attributes it has extracted. It doesn't use the IP address, nor does it validate the address. It just harvests it and passes it on. As far as it is concerned, there are no valid or invalid IPv4 addresses; just 32 bits of arbitrary data.

Negative testing is useful if (1) the UUT performs error handling, or (2) the UUT has an error path that is triggered by one of its direct dependencies. In the first instance (1), the test case can supply a known-bad value and verify that the UUT takes the appropriate action. In the second instance (2), the test case can supply a _test double_ that causes the UUT to take the error path and verify that it has done so.

Inspecting the coverage report for this test shows us that there are no unexercised paths related to IP address values. There is therefore no value in testing for an invalid IP address.

There are, however, untaken paths related to receiving IPv6 addresses and to `switchlink_get_ecmp_info()` returning NOT_FOUND status. These _are_ candidates for additional test cases.